javalib/src/main/scala/java/util/ArrayList.scala

```diff
@@ -14,80 +14,186 @@ package java.util
   }
 
   def ensureCapacity(minCapacity: Int): Unit = {
-    // We ignore this as js.Array doesn't support explicit pre-allocation
+    if (isWebAssembly) {
+      if (innerWasm.length < minCapacity) {
+        if (minCapacity > 1073741824) { // 2^30
+          resizeTo(minCapacity)
+        } else {
+          var powerOf2: Int = 1
+          while (powerOf2 < minCapacity)
+            powerOf2 *= 2
+          resizeTo(powerOf2)
+        }
+      }
+    }
+    // We ignore this in JS as js.Array doesn't support explicit pre-allocation
   }
 
   def size(): Int =
-    inner.length
-
-  override def clone(): AnyRef =
-    new ArrayList(inner.jsSlice(0))
+    if (isWebAssembly) _size
+    else innerJS.length
+
+  override def clone(): AnyRef = {
+    if (isWebAssembly)
+      new ArrayList(innerWasm.clone(), size())
+    else
+      new ArrayList(innerJS.jsSlice(0), 0)
+  }
 
   def get(index: Int): E = {
     checkIndexInBounds(index)
-    inner(index)
+    if (isWebAssembly)
+      innerWasm(index).asInstanceOf[E]
+    else
+      innerJS(index)
   }
 
   override def set(index: Int, element: E): E = {
     val e = get(index)
-    inner(index) = element
+    if (isWebAssembly)
+      innerWasm(index) = element.asInstanceOf[AnyRef]
+    else
+      innerJS(index) = element
     e
   }
 
   override def add(e: E): Boolean = {
-    inner.push(e)
+    if (isWebAssembly) {
+      if (size() >= innerWasm.length)
+        expand()
+      innerWasm(size()) = e.asInstanceOf[AnyRef]
+      _size += 1
+    } else {
+      innerJS.push(e)
+    }
     true
   }
 
   override def add(index: Int, element: E): Unit = {
     checkIndexOnBounds(index)
-    inner.splice(index, 0, element)
+    if (isWebAssembly) {
+      if (size() >= innerWasm.length)
+        expand()
+      System.arraycopy(innerWasm, index, innerWasm, index + 1, size() - index)
+      innerWasm(index) = element.asInstanceOf[AnyRef]
+      _size += 1
+    } else {
+      innerJS.splice(index, 0, element)
+    }
   }
 
   override def remove(index: Int): E = {
     checkIndexInBounds(index)
-    arrayRemoveAndGet(inner, index)
+    if (isWebAssembly) {
+      val removed = innerWasm(index).asInstanceOf[E]
+      System.arraycopy(innerWasm, index + 1, innerWasm, index, size() - index - 1)
+      innerWasm(size - 1) = null // free reference for GC
+      _size -= 1
+      removed
+    } else {
+      arrayRemoveAndGet(innerJS, index)
+    }
   }
 
   override def clear(): Unit =
-    inner.length = 0
+    if (isWebAssembly) {
+      Arrays.fill(innerWasm, null) // free references for GC
+      _size = 0
+    } else {
+      innerJS.length = 0
+    }
 
   override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
     c match {
       case other: ArrayList[_] =>
-        inner.splice(index, 0, other.inner.toSeq: _*)
+        checkIndexOnBounds(index)
+        if (isWebAssembly) {
+          if (size() + other.size() >= innerWasm.length)
+            resizeTo(size() + other.size())
```

Comment:
Let's use `ensureCapacity` here as well, so we keep powers of 2 everywhere.
```suggestion
          ensureCapacity(size() + other.size())
```